js_modules/dagster-ui/packages/ui-core/src/runs/useCursorAccumulatedQuery.tsx

```diff
@@ -1,45 +1,125 @@
-import {useQuery} from '@apollo/client';
+import {OperationVariables, useApolloClient} from '@apollo/client';
 import {DocumentNode} from 'graphql';
-import {useEffect, useState} from 'react';
+import {useEffect, useMemo, useState} from 'react';
 
-import {useQueryRefreshAtInterval} from '../app/QueryRefresh';
+import {useRefreshAtInterval} from '../app/QueryRefresh';
 
-type FetchState = {
+type FetchResult<DataType, CursorType, ErrorType> = {
+  data: DataType[];
   hasMore: boolean;
-  cursor: string | null | undefined;
+  cursor: CursorType | undefined;
+  error: ErrorType | undefined;
 };
 
-export function useCursorAccumulatedQuery<T, TVars extends {cursor?: string | null}, U>(options: {
+type FetcherFunction<DataType, CursorType, ErrorType> = (
+  cursor: CursorType | undefined,
+) => Promise<FetchResult<DataType, CursorType, ErrorType>>;
+
+class AccumulatingDataFetcher<DataType, CursorType, ErrorType> {
+  private fetchData: FetcherFunction<DataType, CursorType, ErrorType>;
+  private onData: (data: DataType[]) => void;
+  private onError?: (error: ErrorType) => void;
+
+  private hasMoreData = true;
+  private dataSoFar: DataType[] = [];
+  private currentCursor: CursorType | undefined = undefined;
+  private fetchPromise?: Promise<void>;
+  private stopped: boolean = false;
+
+  constructor({
+    fetchData,
+    onData,
+    onError,
+  }: {
+    fetchData: FetcherFunction<DataType, CursorType, ErrorType>;
+    onData: (data: DataType[]) => void;
+    onError?: (error: ErrorType) => void;
+  }) {
+    this.fetchData = fetchData;
+    this.onData = onData;
+    this.onError = onError;
+  }
+
+  fetch = async () => {
+    if (this.fetchPromise) {
+      return await this.fetchPromise;
+    }
+    this.fetchPromise = new Promise(async (res) => {
+      // make at least one request
+      this.hasMoreData = true;
+
+      // continue requesting with updated cursors + accumulating data until
+      // stop() is called or hasMore=false.
+      while (this.hasMoreData && !this.stopped) {
+        const {cursor, hasMore, data, error} = await this.fetchData(this.currentCursor);
+        if (this.stopped) {
+          break;
+        }
+        if (error) {
+          this.onError?.(error);
+          break;
+        }
+        this.currentCursor = cursor;
+        this.hasMoreData = hasMore;
+        if (data.length > 0) {
+          this.dataSoFar = this.dataSoFar.concat(data);
+          this.onData(this.dataSoFar);
+        }
+      }
+      res();
+    });
+    const result = await this.fetchPromise!;
+    this.fetchPromise = undefined;
+    return result;
+  };
+
+  stop = () => {
```

Comment:
Had to make these bound methods so we can destructure / pass them to `useRefreshAtInterval`